projects/fluid_simulation/sim.cpp

```diff
@@ -415,8 +421,128 @@ void FluidSimulation::DrawDisplay()
     uint32_t    height      = GetApp()->GetWindowHeight();
     ppx::float2 texelSize   = ppx::float2(1.0f / width, 1.0f / height);
     ppx::float2 ditherScale = mDitheringTexture->GetDitherScale(width, height);
-    ScheduleDR(mDisplayMaterial->GetDR(mDyeTexture[0].get(), mBloomTexture.get(), mSunraysTexture.get(), mDitheringTexture.get(), mDisplayTexture.get(), texelSize, ditherScale));
-    ScheduleDR(mDraw->GetDR(mDisplayTexture.get()));
+    DrawColor(NormalizeColor(GetConfig().backColor));
+    ScheduleDR(mDisplay->GetDR(mDyeTexture[0].get(), mBloomTexture.get(), mSunraysTexture.get(), mDitheringTexture.get(), mDisplayTexture.get(), texelSize, ditherScale));
+    ScheduleDR(mDraw->GetDR(mDisplayTexture.get(), ppx::float2(-1.0f, 1.0f)));
+}
+
+void FluidSimulation::DrawTextures()
+{
+    std::vector<Texture*> v = {
+        mBloomTexture.get(),
+        mCurlTexture.get(),
+        mDivergenceTexture.get(),
+        mPressureTexture[0].get(),
+        mPressureTexture[1].get(),
+        mVelocityTexture[0].get(),
+        mVelocityTexture[1].get(),
+    };
+    auto  coord   = ppx::float2(-1.0f, 1.0f);
+    float maxDimY = 0.0f;
+    for (const auto& t : v) {
+        auto dim = t->GetNormalizedSize();
+        if (coord.x + dim.x >= 1.0) {
+            coord.x = -1.0;
+            coord.y -= maxDimY;
+            maxDimY = 0.0f;
+        }
+        PPX_LOG_DEBUG("Scheduling texture draw for " << t->GetName() << " with normalized dimensions " << t->GetNormalizedSize() << " at coordinate " << coord << "\n");
+        ScheduleDR(mDraw->GetDR(t, coord));
+        coord.x += dim.x + 0.005f;
+        if (dim.y > maxDimY) {
+            maxDimY = dim.y + 0.005f;
+        }
+    }
+}
+
+void FluidSimulation::Update()
+{
+    float delta = CalcDeltaTime();
+
+    // If the marble has been selected, move it around and drop it at random.
+    if (GetConfig().marble) {
+        MoveMarble();
+
+        // Update the color of the marble.
+        if (mRandom.Float() <= GetConfig().colorUpdateFrequency) {
+            mMarble.color = GenerateColor();
+        }
+
+        // Drop the marble at random.
+        if (mRandom.Float() <= GetConfig().marbleDropFrequency) {
+            ppx::float2 delta = mMarble.delta * GetConfig().splatForce;
+            Splat(mMarble.coord, delta, mMarble.color);
+        }
+    }
+
+    // Queue up some splats at random. But limit the amount of outstanding splats so it doesn't get too busy.
+    if (Random().Float() <= GetConfig().splatFrequency) {
+        MultipleSplats(1);
+    }
+
+    Step(delta);
+    Render();
+}
+
+float FluidSimulation::CalcDeltaTime()
```

Comment:
It would be nice for this sample to handle `--deterministic`, so we could use it for screendiff tests.

Reply:
Yes.  I was thinking about `--deterministic` recently and thought that maybe it may be enough with giving  `ppx::Random` a fixed seed when `--deterministic` is used.  WDYT?

Reply:
yes, seeding rand to 0 is fine. But time-based effect should also be avoided.
ppx::Application has `GetElapsedSeconds` which returns the frame count when `--deterministic` is passed. But timers don't.

Reply:
AFrame-count / 60.f actually, since we assume 60fps.